src/block/block_ckpt.c

```diff
@@ -447,18 +450,50 @@ __ckpt_add_blkmod_entry(
     /* Make sure we're not going to run past the end of the bitmap */
     WT_ASSERT(session, blk_mod->bitstring.size >= __bitstr_size((uint32_t)blk_mod->nbits));
     WT_ASSERT(session, end_bit < blk_mod->nbits);
-    /* Set all the bits needed to record this offset/length pair. */
-    __bit_nset(blk_mod->bitstring.mem, start_bit, end_bit);
+    /* Change all the bits needed to record this offset/length pair. */
+    if (set)
+        __bit_nset(blk_mod->bitstring.mem, start_bit, end_bit);
+    else {
+        /*
+         * We can only clear full ranges represented by bits. Ignore any partial ranges at the
+         * beginning and end of the offset/length range but clear any full bit ranges in between.
+         */
+        adj = (uint64_t)offset % gran;
+        /*
+         * Adjust partial ranges at the beginning and the end of the offset/length range. Round up
+         * the offset to the next granularity boundary.
+         */
+        if (adj != 0)
+            adj = gran - adj;
+        clr_off = offset + (wt_off_t)adj;
+        /*
+         * Deduct any initial partial length from the overall length. Then round it down with
+         * integer division to a granularity multiple.
+         */
+        clr_len = ((len - (wt_off_t)adj) / (wt_off_t)gran) * (wt_off_t)gran;
+        WT_ASSERT(session, clr_off % (wt_off_t)gran == 0);
+        WT_ASSERT(session, clr_len % (wt_off_t)gran == 0);
+        if (clr_len > 0) {
+            /* Calculate the new full bits we can clear. */
+            start_bit = (uint64_t)clr_off / gran;
+            end_bit = (uint64_t)(clr_off + clr_len - 1) / gran;
+            WT_ASSERT(session, end_bit >= start_bit);
+            __wt_verbose(session, WT_VERB_TEMPORARY,
+              "MOD_BLKMOD: %s: %s: gran %d offset %d len %d, adj off %d len %d, %s bits %d-%d",
+              blk_mod->id_str, msg, (int)gran, (int)offset, (int)len, (int)clr_off, (int)clr_len,
+              set ? "SET" : "CLEAR", (int)start_bit, (int)end_bit);
+            __bit_nclr(blk_mod->bitstring.mem, start_bit, end_bit);
+        }
+    }
     return (0);
 }
 
 /*
- * __ckpt_add_blk_mods_alloc --
+ * __ckpt_live_blkmods --
  *     Add the checkpoint's allocated blocks to all valid incremental backup source identifiers.
```

Comment:
This comment needs to be updated to reflect the fact that the function now processes the discard list, too.

Reply:
Fixed